numpyro/util.py

```diff
@@ -298,3 +298,29 @@ def unravel_pytree(arr):
         return tree_unflatten(treedef, unravel_list(arr))
 
     return flat, unravel_pytree
+
+
+def soft_vmap(fn, xs, batch_shape, chunk_size=None):
```

Comment:
@fehiepsi could you please add a docstring for this? i'd like to better understand what it's doing...

Reply:
Sure, I'll add. This basically computes `fn(xs)` by vectorizing `fn(x)` over `batch_shape` (i.e. `xs.shape = batch_shape + 'x.shape'`). The optional `chunk_size` allows the job taken in chunk: by spliting `xs` into chunks `xs_1`,..., `xs_n`, computing `fn(xs_1)`, ... `fn(xs_n)` using `vmap`, and finally merging the results. `chunk_size=None` means that there is only 1 chunk, which is the same as `vmap(fn)(x)`. `chunk_size=1` means that we don't use `vmap` at all, which is the same as `map(fn)(x)`.

The difference w.r.t. your script is `batch_shape` argument, which has length 1 in your script but has arbitrary length here (to allow calculating `num_chains x num_samples x ...`, which has length 2, forms).

Reply:
I just add the docstring.